lighthouse-core/test/gather/driver-test.js

```diff
@@ -482,60 +482,70 @@ describe('.gotoURL', () => {
       .mockResponse('Runtime.evaluate');
   });
 
-  it('will track redirects through gotoURL load', async () => {
-    const delay = () => new Promise(resolve => setTimeout(resolve));
+  for (const finalUrlSource of ['network', 'window.location']) {
+    it(`will track redirects through gotoURL via ${finalUrlSource}`, async () => {
+      const delay = () => new Promise(resolve => setTimeout(resolve));
 
-    class ReplayConnection extends Connection {
-      connect() {
-        return Promise.resolve();
-      }
-      disconnect() {
-        return Promise.resolve();
-      }
-      replayLog() {
-        redirectDevtoolsLog.forEach(msg => this.emitProtocolEvent(msg));
-      }
-      /**
-       * @param {string} method
-       * @param {any} _
-       */
-      sendCommand(method, _) {
-        const resolve = Promise.resolve();
-
-        // If navigating, wait, then replay devtools log in parallel to resolve.
-        if (method === 'Page.navigate') {
-          resolve.then(delay).then(_ => this.replayLog());
+      class ReplayConnection extends Connection {
+        connect() {
+          return Promise.resolve();
+        }
+        disconnect() {
+          return Promise.resolve();
+        }
+        replayLog() {
+          redirectDevtoolsLog.forEach(msg => this.emitProtocolEvent(msg));
+        }
+        /**
+         * @param {string} method
+         * @param {any} _
+         */
+        sendCommand(method, _) {
+          const resolve = Promise.resolve();
+
+          // If navigating, wait, then replay devtools log in parallel to resolve.
+          if (method === 'Page.navigate') {
+            resolve.then(delay).then(_ => this.replayLog());
+          }
+
+          return resolve;
         }
-
-        return resolve;
       }
-    }
-    const replayConnection = new ReplayConnection();
-
-    const driver = /** @type {TestDriver} */ (new Driver(replayConnection));
-
-    // Redirect in log will go through
-    const startUrl = 'http://en.wikipedia.org/';
-    // then https://en.wikipedia.org/
-    // then https://en.wikipedia.org/wiki/Main_Page
-    const finalUrl = 'https://en.m.wikipedia.org/wiki/Main_Page';
-
-    const loadOptions = {
-      waitForLoad: true,
-      passContext: {
-        passConfig: {
-          pauseAfterLoadMs: 0,
-          networkQuietThresholdMs: 0,
-          cpuQuietThresholdMs: 0,
+      const replayConnection = new ReplayConnection();
+
+      const driver = /** @type {TestDriver} */ (new Driver(replayConnection));
+      const evalauteAsync = driver.evaluateAsync.bind(driver);
+      driver.evaluateAsync = (expression, options) => {
```

Comment:
was wondering if this could be done as a mockResponse, but I guess this fine too?

Reply:
this test uses its own mock connection already so it'd be kindof tricky and this seemed pretty straightforward compared to the mix/match of connection delegation